### PR TITLE
 [Bug] Using default animation for errors that occur.

### PR DESCRIPTION
**Pull request by podarsmarty**
## What are the changes the user will see?
- Fetch errors will no longer result in a crashed game

## Why am I making these changes?
- These errors are prevalent in offline builds that crash the game. I expect this to be possible on the live version but more rare and transient

## What are the changes from a developer perspective?
- Adds error handling to promise so that default animations do not occur ONLY for bad responses

### Screenshots/Videos
N/A

## How to test the changes?
- Build and package offline with move animation json file for a pokemon being used deleted. Will crash before and work properly after

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I considered writing automated tests for the issue?
     I took a look but for tests the entirely of fetch is mocked out. would need to add mock configurations to add unit tests
- [ ] If I have text, did I make it translatable and add a key in the English locale file(s)?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [ ] Are the changes visual?
  - [ ] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: src/data/battle-anims.ts [hunk 1]
```diff
@@ -515,13 +514,23 @@ export function initMoveAnim(scene: BattleScene, move: Moves): Promise<void> {
...
+            useDefaultAnim(move, defaultMoveAnim, error);
+            return resolve();
           });
       };
       fetchAnimAndResolve(move);
     }
   });
 }
 
+function useDefaultAnim(move: Moves, defaultMoveAnim: Moves, ...optionalParams: any[]) {
```
Comment (flx-sta) [inline src/data/battle-anims.ts:528]: NIT:
missing ts-docs

File: src/enums/moves.ts [hunk 2]
```diff
@@ -1873,3 +1873,9 @@ export enum Moves {
   /**{@link https://bulbapedia.bulbagarden.net/wiki/Malignant_Chain_(move) | Source} */
   MALIGNANT_CHAIN,
 }
+
+export class MovesUtil {
```
Comment (flx-sta) [inline src/enums/moves.ts:1877]: NIT:
Missing ts-docs

File: src/enums/moves.ts [hunk 3]
```diff
@@ -1873,3 +1873,9 @@ export enum Moves {
...
+  public static filename(move: Moves) {
```
Comment (flx-sta) [inline src/enums/moves.ts:1878]: NIT:
Missing ts-docs

File: src/enums/moves.ts (hunk 2 above)
Comment (flx-sta) [reply src/enums/moves.ts:1877]: Also I'd prefer to move new classes into their own files.

To keep it simple though you might want to consider simply creating an exported function/method in `utils.ts`

**Review by flx-sta — CHANGES_REQUESTED** [review]

File: src/data/battle-anims.ts [hunk 4]
```diff
@@ -515,13 +514,30 @@ export function initMoveAnim(scene: BattleScene, move: Moves): Promise<void> {
             } else {
               resolve();
             }
+          })
+          .catch(error => {
+            useDefaultAnim(move, defaultMoveAnim, error);
+            return resolve();
```
Comment (flx-sta) [inline src/data/battle-anims.ts:520]: We should probably print at least a log, better yet a warning here
```suggestion
          .catch(error => {
            console.warn(`No move animation found for ${Utils.animationFileName(move)}! Falling back to default.`);
            useDefaultAnim(move, defaultMoveAnim, error);
            return resolve();
```

**Review by flx-sta — CHANGES_REQUESTED** [review]
sorry found 1 more.
(make sure to not always force push 👀 )

**Comment by podarsmarty** [comment]
> sorry found 1 more. (make sure to not always force push 👀 )

ahh no worries. will fix. my bad, saw you were merging, figured id rebase on beta and squash my commits so forced

File: src/data/battle-anims.ts (hunk 4 above)
Comment (podarsmarty) [reply src/data/battle-anims.ts:520]: @flx-sta hmm the useDefaultAnim function already logs an error to the console

Comment (podarsmarty) [reply src/data/battle-anims.ts]: renamed method to make that behavior more clear

**Comment by flx-sta** [comment]
I find it rather odd that you decided to move the logging into the defaultAnim function (although you did rename it).
But technically the functions (original) context has nothing to with the logging. It would be better to have the logging inside the actual `catch()`.
Also `console.error` seems to elevated as it's not breaking anything. It's just an inconvenience, a warning to say so. Which implies wanting to use `console.warn()`

**Comment by podarsmarty** [comment]
> I find it rather odd that you decided to move the logging into the defaultAnim function (although you did rename it). But technically the functions (original) context has nothing to with the logging. It would be better to have the logging inside the actual `catch()`. Look at it this way: Technically you could be calling `defaultAnim()` on purpose without it being an error/warning, right?
> 
> Also `console.error` seems to elevated as it's not breaking anything. It's just an inconvenience, a warning to say so. Which implies wanting to use `console.warn()`

i had moved it since the function is called in two places that would both warrant a log. figured it be better to share rather than copy pasting it. yea its not breaking, i had left it as an error the way it originally was, i can switch it over to a warn and duplicate the logging though.

however, i'd argue you should only be calling defaultAnim() in an error state since everything should have an animation defined the way it is setup

**Review by flx-sta — DISMISSED** [review]
Thanks for elaborating your approach.

File: src/data/battle-anims.ts [hunk 5]
```diff
@@ -515,13 +515,41 @@ export function initMoveAnim(scene: BattleScene, move: Moves): Promise<void> {
...
       fetchAnimAndResolve(move);
     }
   });
 }
 
+/**
+ * Populates the default animation for the given move.
+ *
+ * @param move the move to populate an animation for
+ * @param defaultMoveAnim the default move to use as the default animation
```
Comment (DayKev) [inline src/data/battle-anims.ts:534]:
```suggestion
 * @param defaultMoveAnim the move to use as the default animation
```

File: src/data/battle-anims.ts [hunk 6]
```diff
@@ -515,13 +515,41 @@ export function initMoveAnim(scene: BattleScene, move: Moves): Promise<void> {
...
+ */
+function useDefaultAnim(move: Moves, defaultMoveAnim: Moves) {
+  populateMoveAnim(move, moveAnims.get(defaultMoveAnim));
+}
+
+/**
+ * Helper method to call when an move animation is missing.
```
Comment (DayKev) [inline src/data/battle-anims.ts:541]:
```suggestion
 * Helper method for printing a warning to the console when a move animation is missing.
```

File: src/data/battle-anims.ts [hunk 7]
```diff
@@ -515,13 +515,41 @@ export function initMoveAnim(scene: BattleScene, move: Moves): Promise<void> {
...
+ *
+ * @param move the move to populate an animation for
+ * @param optionalParams parameters to add to the error logging
+ *
+ * @remarks use {@link useDefaultAnim} to use a default animation
```
Comment (DayKev) [inline src/data/battle-anims.ts:546]:
```suggestion
 * @remarks use {@linkcode useDefaultAnim} to use a default animation
```

**Review by flx-sta — APPROVED** [review]

**Review by DayKev — APPROVED** [review]

**Pull request merged**